### PR TITLE
Generalize bound type for `BoundedVec<T, S>` #11076

### PR DESCRIPTION
**Pull request by jorgeantonio21**
Issue #11076 

Current progress: 

1. Refactor BoundedVec implementation to add a second phantom type B and refactor other necessary bits of code.

2. Created a AtMost32BitUnsigned trait (in `primitives/arithmetic/src/traits.rs`), with implementations for u8, u16 and u32. It is probably cleaner to add a macro to deal with this directly.

3. The code does compile and tests were added to cover new functionality on the bound type for `BoundedVec<T, S>`

### CODE REVIEW

**Comment by bkchr** [comment]
For what do we need this?

**Comment by jorgeantonio21** [comment]
> For what do we need this?

It is a work in progress to tackle the issue #11076, toallow the `bound` type on BoundedVec<T, S> to either one of u8, u16, u32 primitive types.

**Comment by kianenigma** [comment]
@bkchr @KiChjang: @jorgeantonio21 is a hiring candidate and we're using this issue as an evaluation task. We can wear our boxing gloves later and discuss if this change should be merged or not, but let's continue working on it for the time being.

File: frame/support/src/storage/bounded_vec.rs [hunk 1]
```diff
@@ -22,8 +22,13 @@ use crate::{
 	storage::{StorageDecodeLength, StorageTryAppend},
 	traits::{Get, TryCollect},
 	WeakBoundedVec,
+	
 };
+
+extern crate sp_arithmetic;
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:28]: this syntax is deprecated.

File: frame/support/Cargo.toml [hunk 2]
```diff
@@ -28,6 +28,7 @@ sp-staking = { version = "4.0.0-dev", default-features = false, path = "../../pr
 tt-call = "1.0.8"
 frame-support-procedural = { version = "4.0.0-dev", default-features = false, path = "./procedural" }
 paste = "1.0"
+funty = "2.0.0"
```
Comment (kianenigma) [inline frame/support/Cargo.toml:31]: why is this needed?

File: frame/support/src/storage/bounded_vec.rs [hunk 3]
```diff
@@ -39,60 +44,60 @@ use sp_std::{marker::PhantomData, prelude::*};
...
-pub struct BoundedVec<T, S>(Vec<T>, PhantomData<S>);
+pub struct BoundedVec<T, B, S: Get<B>>(Vec<T>, PhantomData<B>, PhantomData<S>);
 
 /// A bounded slice.
 ///
 /// Similar to a `BoundedVec`, but not owned and cannot be decoded.
 #[derive(Encode, scale_info::TypeInfo)]
 #[scale_info(skip_type_params(S))]
-pub struct BoundedSlice<'a, T, S>(&'a [T], PhantomData<S>);
+pub struct BoundedSlice<'a, T, B, S: Get<B>>(&'a [T], PhantomData<B>, PhantomData<S>);
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:54]: similarly

File: frame/support/src/storage/bounded_vec.rs [hunk 4]
```diff
@@ -39,60 +44,60 @@ use sp_std::{marker::PhantomData, prelude::*};
...
 }
 
-impl<'a, T, S: Get<u32>> TryFrom<&'a [T]> for BoundedSlice<'a, T, S> {
+impl<'a, T, B: AtMost32BitUnsigned, S: Get<B>> TryFrom<&'a [T]> for BoundedSlice<'a, T, B, S> {
 	type Error = ();
 	fn try_from(t: &'a [T]) -> Result<Self, Self::Error> {
-		if t.len() < S::get() as usize {
-			Ok(BoundedSlice(t, PhantomData))
+		if S::get().into().try_into().map_or(false, |v: usize| t.len() < v) {
+			Ok(BoundedSlice(t, PhantomData, PhantomData))
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:81]:
```suggestion
			Ok(BoundedSlice(t, PhantomData))
```

File: frame/support/src/storage/bounded_vec.rs [hunk 5]
```diff
@@ -39,60 +44,60 @@ use sp_std::{marker::PhantomData, prelude::*};
...
-impl<T: Decode, S: Get<u32>> Decode for BoundedVec<T, S> {
+impl<T: Decode, B: AtMost32BitUnsigned, S: Get<B>> Decode for BoundedVec<T, B, S> {
 	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
 		let inner = Vec::<T>::decode(input)?;
-		if inner.len() > S::get() as usize {
+		if S::get().into().try_into().map_or(false, |v: usize| inner.len() > v) {
 			return Err("BoundedVec exceeds its limit".into())
 		}
-		Ok(Self(inner, PhantomData))
+		Ok(Self(inner, PhantomData, PhantomData))
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:100]:
```suggestion
		Ok(Self(inner, PhantomData))
```

File: frame/support/src/storage/bounded_vec.rs [hunk 6]
```diff
@@ -101,12 +106,12 @@ impl<T: Decode, S: Get<u32>> Decode for BoundedVec<T, S> {
...
 // `BoundedVec`s encode to something which will always decode as a `Vec`.
-impl<T: Encode + Decode, S: Get<u32>> EncodeLike<Vec<T>> for BoundedVec<T, S> {}
+impl<T: Encode + Decode, B: AtMost32BitUnsigned, S: Get<B>> EncodeLike<Vec<T>> for BoundedVec<T, B, S> {}
 
-impl<T, S> BoundedVec<T, S> {
+impl<T, B, S: Get<B>> BoundedVec<T, B, S> {
 	/// Create `Self` from `t` without any checks.
 	fn unchecked_from(t: Vec<T>) -> Self {
-		Self(t, Default::default())
+		Self(t, Default::default(), Default::default())
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:114]:
```suggestion
		Self(t, Default::default())
```

File: frame/support/src/storage/bounded_vec.rs [hunk 7]
```diff
@@ -203,19 +208,19 @@ impl<T, S> BoundedVec<T, S> {
...
 
-impl<T, S: Get<u32>> BoundedVec<T, S> {
+impl<T, B: AtMost32BitUnsigned, S: Get<B>> BoundedVec<T, B, S> {
 	/// Pre-allocate `capacity` items in self.
 	///
 	/// If `capacity` is greater than [`Self::bound`], then the minimum of the two is used.
 	pub fn with_bounded_capacity(capacity: usize) -> Self {
 		let capacity = capacity.min(Self::bound());
-		Self(Vec::with_capacity(capacity), Default::default())
+		Self(Vec::with_capacity(capacity), Default::default(), Default::default())
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:223]:
```suggestion
		Self(Vec::with_capacity(capacity), Default::default())
```

File: frame/support/src/storage/bounded_vec.rs [hunk 8]
```diff
@@ -225,7 +230,8 @@ impl<T, S: Get<u32>> BoundedVec<T, S> {
 
 	/// Get the bound of the type in `usize`.
 	pub fn bound() -> usize {
-		S::get() as usize
+		let bound: usize = S::get().into().try_into().unwrap();
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:233]: you can never `unwrap` in the runtime, this needs to be made into a safe conversion.

You can go to `u32` and then since wasm is a 32 bit system, you can do `as usize`.

File: frame/support/src/storage/bounded_vec.rs [hunk 9]
```diff
@@ -528,97 +535,155 @@ where
...
 	crate::generate_storage_alias! {
 		Prefix,
-		FooDoubleMap => DoubleMap<(Twox128, u32), (Twox128, u32), BoundedVec<u32, ConstU32<7>>>
+		FooDoubleMap => DoubleMap<(Twox128, u32), (Twox128, u32), BoundedVec<u32, u32, ConstU32<7>>>
 	}
 
 	#[test]
 	fn slide_works() {
-		let mut b: BoundedVec<u32, ConstU32<6>> = bounded_vec![0, 1, 2, 3, 4, 5];
+		let mut b: BoundedVec<u32, u32, ConstU32<6>> = bounded_vec![0, 1, 2, 3, 4, 5];
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:632]: what is the rationale for making _all_ tests work with all 3 possible bounds now? I would just add one tests that ensures different bound types can be used, and leave all the rest unchanged. 

But you can elaborate more on what scenario you are testing for, perhaps I am missing something.

File: primitives/arithmetic/src/traits.rs [hunk 10]
```diff
@@ -26,6 +26,8 @@ pub use num_traits::{
 use sp_std::ops::{
 	Add, AddAssign, Div, DivAssign, Mul, MulAssign, Rem, RemAssign, Shl, Shr, Sub, SubAssign,
 };
+use funty::AtMost32;
```
Comment (kianenigma) [inline primitives/arithmetic/src/traits.rs:29]: This is easy enough to make in-house, why add a dependency for it? 

Generally, we try to limit dependencies in the runtime and avoid adding new ones unless if really needed.

Comment (jorgeantonio21) [reply primitives/arithmetic/src/traits.rs:29]: Your point makes complete sense, @kianenigma. I think, we don't even need to have a `AtMost32` trait bound, `Into<u32> + Unsigned` should do the job.

File: frame/support/src/storage/bounded_vec.rs (hunk 5 above)
Comment (jorgeantonio21) [reply frame/support/src/storage/bounded_vec.rs:98]: Replacing `Ok(Self(inner, PhantomData, PhantomData))` with `Ok(Self(inner, PhantomData))` gives me a compiler error of

`expected 3 arguments, found 2`,

I think I haven't understood the suggestion. Can we somehow drop the number of type parameters using this approach ?

File: frame/support/src/storage/bounded_vec.rs (hunk 3 above)
Comment (kianenigma) [reply frame/support/src/storage/bounded_vec.rs:52]: There's no need to have two phantom datas, you can wrap multiple types in one with a tuple, like `PhantomData<(A, B, C)>`

Comment (jorgeantonio21) [reply frame/support/src/storage/bounded_vec.rs]: I understand now. This allows to settle the other suggestions as well. Thank you !

File: frame/support/src/storage/bounded_vec.rs (hunk 5 above)
Comment (jorgeantonio21) [reply frame/support/src/storage/bounded_vec.rs]: Issue solved

File: frame/support/src/storage/bounded_btree_map.rs [hunk 11]
```diff
@@ -24,6 +24,9 @@ use crate::{
 use codec::{Decode, Encode, MaxEncodedLen};
 use sp_std::{borrow::Borrow, collections::btree_map::BTreeMap, marker::PhantomData, ops::Deref};
 
+extern crate sp_arithmetic;
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_map.rs:27]:
```suggestion
```

File: primitives/arithmetic/src/traits.rs [hunk 12]
```diff
@@ -153,6 +153,12 @@ pub trait AtLeast32BitUnsigned: AtLeast32Bit + Unsigned {}
 
 impl<T: AtLeast32Bit + Unsigned> AtLeast32BitUnsigned for T {}
 
+// A meta trait for arithmetic. Only allows for unsigned integers, at most of 32 bits, the opposite of [`AtLeast32BitUnsigned`]
+pub trait AtMost32BitUnsigned: Into<u32> + Unsigned {}
+
+impl<T: Into<u32> + Unsigned> AtMost32BitUnsigned for T {}
```
Comment (kianenigma) [inline primitives/arithmetic/src/traits.rs:159]: This has the same issue as you have explained elsewhere: A lot of things are Into<u32>, that are not numbers. Should only and manually be implemented for u8, u16 and u32.

Comment (jorgeantonio21) [reply primitives/arithmetic/src/traits.rs:159]: Even if you make `Into<u32> + Unsigned`? The `Unsigned` trait also impls `Num`, isn't this sufficient to guarantee that we are dealing with 'unsigned numbers'?

**Comment by kianenigma** [comment]
closing as the underlying issue was not really sensible, my bad, and thanks @jorgeantonio21
